gdal/gcore/overview.cpp

```diff
@@ -5049,28 +5049,20 @@ GDALRegenerateOverviewsMultiBand( int nBands, GDALRasterBand** papoSrcBands,
                         apaChunk[iBand],
                         nChunkXSizeQueried, nChunkYSizeQueried,
                         eWrkDataType, 0, 0, nullptr );
-                }
 
-                if( bUseNoDataMask && eErr == CE_None )
-                {
-                    GDALRasterBand* poSrcBand = nullptr;
-                    if( iSrcOverview == -1 )
-                        poSrcBand = papoSrcBands[0];
-                    else
-                        poSrcBand = papapoOverviewBands[0][iSrcOverview];
-                    auto poMaskBand = bIsMask ? poSrcBand : poSrcBand->GetMaskBand();
-                    eErr = poMaskBand->RasterIO(
-                        GF_Read,
-                        nChunkXOffQueried, nChunkYOffQueried,
-                        nChunkXSizeQueried, nChunkYSizeQueried,
-                        pabyChunkNoDataMask,
-                        nChunkXSizeQueried, nChunkYSizeQueried,
-                        GDT_Byte, 0, 0, nullptr );
+                    if( bUseNoDataMask && eErr == CE_None )
+                    {
+                        auto poMaskBand = bIsMask ? poSrcBand : poSrcBand->GetMaskBand();
```

Comment:
could you change that logic to something like
```suggestion
                        auto poMaskBand = bIsMask || poSrcBand->GetColorInterpretation() == GCI_AlphaBand ? poSrcBand : poSrcBand->GetMaskBand();
```
I guess this should lead to the previous result. The fact that we are rough on the edges between alpha=0 and != 0 was a wanted behaviour to avoid "expending" the validity area when computing overviews.

Reply:
Doing that the test ovr_32 leads to the previous result (well, using `papoSrcBands[iBand]` instead of `poSrcBand`. Looks like the color interpretation is not well defined in `papapoOverviewBands`... and I think I had the same symptoms last week in COG).

If I understand correctly, it is using the alpha band itself to mask the alpha band. Then combining alpha band (with only 0 and 255) and "nodata value"=-10000 produces that values that are 0 in the main image, become -10000 in the overviews (instead of stay as 0). I am not sure if that is what we want. What about `auto poMaskBand = bIsMask || (!pabHasNoData[iBand] && papoSrcBands[iBand]->GetColorInterpretation() == GCI_AlphaBand) ? poSrcBand : poSrcBand->GetMaskBand();` ?

About the test ovr_32 creating the overviews for `stefan_full_rgba.tif`, the output is really rough with master.
You can compare the first overview in this images:
Master:  
![master](https://user-images.githubusercontent.com/15678366/120693422-4620bd80-c4a9-11eb-83d2-7ea200106677.png)
This branch (without filtering with GetColorInterpretation): 
![branch](https://user-images.githubusercontent.com/15678366/120693464-520c7f80-c4a9-11eb-9281-c01fdaeed74f.png)

Reply:
> What about `auto poMaskBand = bIsMask || (!pabHasNoData[iBand] && papoSrcBands[iBand]->GetColorInterpretation() == GCI_AlphaBand) ? poSrcBand : poSrcBand->GetMaskBand();` ?

sound plausible, although I'm starting to be a bit lost :-) A comment in the code explaining the scenario you mention would be useful. I guess the GetMaskBand() logic could have a special case when both nodata and alpha are present, but I'm not sure we want to get into that level of complication.

> About the test ovr_32 creating the overviews for stefan_full_rgba.tif, the output is really rough with master.

yeah, for a real alpha channel like this image with not just 0 and 255 values, this doesn't look good. The current behavior was rather tuned for an alpha channel that is only 0 and 255, so just masking. Perhaps we could select the appropriate behavior by analyzing the content of the mask if it comes from an alpha band : is it just 0 and 255, or does it include intermediate values ? But that should be done in a separate pull request as I'd like that one to be the minimum to fix the broken case to be backportable.

Reply:
Yes, some things better for a later PR. I squashed the commits in one, including the GetColorInterpretation usage.